tests/unit/labels/MappedLabels_unittest.cc

```diff
@@ -0,0 +1,211 @@
+		auto result = new CBinaryLabels(labs);
+		SG_REF(result);
+		return result;
+	}
+
+	virtual CMulticlassLabels* apply_multiclass(CFeatures*) override
+	{
+		auto labs = SGVector<float64_t>(10);
+		auto num_classes = m_labels->as<CMulticlassLabels>()->get_labels().unique().vlen;
+		ASSERT(num_classes);
+		for (auto i : range(labs.size()))
+			labs[i] = i % num_classes;
+		auto result = new CMulticlassLabels(labs);
+		SG_REF(result);
+		return result;
+	}
+
+	virtual EProblemType get_machine_problem_type() const
+	{
+		return m_problem_type;
+	}
+
+	EProblemType m_problem_type;
+};
+
+class MappedLabelsFixture : public ::testing::Test
+{
+protected:
+	void SetUp()
+	{
+		feats = nullptr;
+		m = nullptr;
+		result = nullptr;
+	}
+
+	void TearDown()
+	{
+		SG_UNREF(feats);
+		SG_UNREF(m);
+		SG_UNREF(result);
+	}
+
+public:
+	CLabels* train_and_apply(CLabels* labs, EProblemType pt)
+	{
+		ASSERT(labs);
+
+		X = SGMatrix<float64_t>(1, labs->get_num_labels());
+		feats = new CDenseFeatures<float64_t>(X);
+		SG_REF(feats);
+
+		m = new MockMachine(pt);
+		SG_REF(m);
+
+		m->set_labels(labs);
+		m->train(feats);
+		result = m->apply();
+		return result;
+	}
+
+	template <class USER_PROVIDED, class MAPPING, class EXPECTED_INTERNAL>
+	void test_mapped_labels(SGVector<float64_t> user_provided, SGVector<float64_t> expected_internal, EProblemType pt)
+	{
+		auto labs = some<USER_PROVIDED>(user_provided);
+		auto applied = train_and_apply(labs, pt);
+
+		// CMachine::apply returns the same class as user provided
+		EXPECT_TRUE(dynamic_cast<USER_PROVIDED*>(applied));
+
+		// internal labels are of appropriate class and mapping works
+		auto internal = m->get_labels();
+		auto casted = dynamic_cast<MAPPING*>(internal);
+		ASSERT_TRUE(casted);
+		EXPECT_TRUE(casted->get_labels().equals(expected_internal));
+
+		// inversion mapping works: mapping of the internal labels gives original
+		auto invert_labels = some<EXPECTED_INTERNAL>(expected_internal);
+		auto inverted = casted->invert(invert_labels);
+		EXPECT_TRUE(inverted->get_labels().equals(labs->get_labels()));
+		SG_UNREF(inverted);
+		SG_UNREF(internal);
+	}
+
+	template <class USER_PROVIDED>
+	void test_mapped_labels(SGVector<float64_t> user_provided, EProblemType pt)
+	{
+		auto labs = some<USER_PROVIDED>(user_provided);
+		auto applied = train_and_apply(labs, pt);
+
+		// CMachine::apply returns the same class as user provided
+		EXPECT_TRUE(dynamic_cast<USER_PROVIDED*>(applied));
+
+		// internal labels are the same instance as user provided
+		auto converted = m->get_labels();
+		EXPECT_TRUE(converted == labs);
+		SG_UNREF(converted);
+	}
+
+	SGMatrix<float64_t> X;
+	CFeatures* feats;
+	CMachine* m;
+	CLabels* result;
+};
+
+TEST_F(MappedLabelsFixture, mc_as_bin)
+{
+	test_mapped_labels<CMulticlassLabels, CMappedBinaryLabels, CBinaryLabels>({0,2,0,2}, {-1,1,-1,1}, PT_BINARY);
+}
+
+TEST_F(MappedLabelsFixture, bin_as_bin)
+{
+	test_mapped_labels<CBinaryLabels>({-1,1,-1,1}, PT_BINARY);
```

Comment:
this ensures that no new instances are created when labels do not need to be mapped (so far this was the case in almost all machines where we asserted the necessary label type)